### PR TITLE
feat(#2264): Expose Retry Handler

### LINKED ISSUE #2264: Retry
Similar to https://github.com/nodejs/undici/issues/2256

Sharing some internal code we have been using in case someone wants to work on it and add it as a util in undici.

```js
const assert = require('node:assert')
const stream = require('node:stream')
const { isReadableNodeStream } = require('../stream')
const { parseHeaders } = require('../http')
const createError = require('http-errors')

function parseRange(range) {
  const m = range ? range.match(/^bytes=(\d+)-(\d+)?$/) : null
  return m ? { start: parseInt(m[1]), end: m[2] ? parseInt(m[2]) : null } : null
}

module.exports = class RetryHandler {
  /** @type {Object} */ opts
  /** @type {Object} */ dispatcher
  /** @type {Error | undefined} */ retryError
  /** @type {Number} */ retryCount = 0
  /** @type {Number} */ retryMax
  /** @type {Array<Number>} */ retryCode
  /** @type {Array<String>} */ retryMessage
  /** @type {Array<Number>} */ retryStatus
  /** @type {Array<String>} */ retryMethod
  /** @type {Boolean} */ idempotent
  /** @type {Number} */ position = 0
  /** @type {NodeJS.Timeout | null} */ timeout
  /** @type {{start: Number, end: Number | null} | null} */ range

  constructor (opts, { dispatcher, handler }) {
    const range = opts.headers?.range ?? opts.headers?.range

    this.dispatcher = dispatcher
    this.handler = handler
    this.opts = opts
    this.statusCode = 0
    this.idempotent = opts.idempotent

    const {
      count: retryMax = 8,
      method: retryMethod = ['GET', 'HEAD', 'OPTIONS', 'PUT', 'DELETE', 'TRACE', 'PATCH'],
      status: retryStatus = [420, 429, 502, 503, 504],
      code: retryCode = [
        'ECONNRESET',
        'ECONNREFUSED',
        'ENOTFOUND',
        'ENETDOWN',
        'ENETUNREACH',
        'EHOSTDOWN',
        'EHOSTUNREACH',
        'EPIPE',
      ],
      message: retryMessage = ['other side closed'],
    } = opts.retry ?? {}

    this.retryMax = retryMax
    this.retryCount = retryCode
    this.retryMessage = retryMessage
    this.retryStatus = retryStatus
    this.retryMethod = retryMethod

    this.range = opts.method === 'GET'
      ? range
        ? parseRange(range)
        : { start: 0, end: null }
      : null
  }

  onConnect(abort) {
    this.handler.onConnect(err => {
      if (this.timeout) {
        clearTimeout(this.timeout)
      }
      abort(err)
    })
  }

  onHeaders(statusCode, rawHeaders, resume, statusMessage) {
    if (statusCode >= 300) {
      throw createError(statusCode, { headers: parseHeaders(rawHeaders) })
    }

    let contentLength
    for (let i = 0; i < rawHeaders.length; i += 2) {
      const key = rawHeaders[i].toString()
      if (key.length === 'content-length'.length && key.toLowerCase() === 'content-length') {
        contentLength = rawHeaders[key]
      }
    }

    if (statusCode === 200 && this.range?.end === null && contentLength) {
      this.range.end = Number(contentLength)
      assert(Number.isFinite(this.range.end), 'invalid content-length')
    }

    if (this.statusCode) {
      assert(this.retryError)

      if (statusCode === 200) {
        // TODO (feat): Resume 200 if distance is not too far?
        throw this.retryError
      } else if (statusCode === 206) {
        // TODO (fix): Check content-range.
        return true // TODO (fix): return false if body is full...
      } else {
        throw this.retryError
      }
    } else {
      this.statusCode = statusCode
      return this.handler.onHeaders(statusCode, rawHeaders, resume, statusMessage)
    }
  }

  onData(chunk) {
    this.position += chunk.length
    return this.handler.onData(chunk)
  }

  onComplete(rawTrailers) {
    return this.handler.onComplete(rawTrailers)
  }

  onError(err) {
    if (this.statusCode) {
      if (this.range && this.retryCount < this.retryMax && this.opts.methid === 'GET') {
        // this.logger?.warn({ err, retryCount: this.retryCount }, 'upstream response retrying')

        this.retryError = err
        this.retryCount += 1

        this.dispatcher.dispatch(
          {
            ...this.opts,
            headers: {
              ...this.opts.headers,
              range: `bytes=${this.range.start + this.position}-${
                this.range.end ? this.range.end : ''
              }`,
            },
          },
          this,
        )
      } else {
        return this.handler.onError(err)
      }
    } else if (
      err.name !== 'AbortError' &&
      this.retryCount < this.retryMax &&
      (this.opts.body == null ||
        typeof this.opts.body === 'string' ||
        Buffer.isBuffer(this.opts.body) ||
        // @ts-ignore: isDisturbed is not in typedefs
        (isReadableNodeStream(this.opts.body) && !stream.isDisturbed(this.opts.body))) &&
      (this.idempotent || this.retryMethod.includes(this.opts.method)) &&
      (this.retryCode.includes(err.code) ||
        this.retryMessage.includes(err.message) ||
        this.retryStatus.includes(err.statusCode))
    ) {
      const delay =
        parseInt(err.headers?.['Retry-After']) * 1e3 ||
        Math.min(10e3, this.retryCount * 1e3 + 1e3)

      // this.logger?.warn({ err, retryCount: this.retryCount, delay }, 'upstream request retrying')

      this.retryError = err
      this.retryCount += 1

      this.timeout = setTimeout(() => {
        this.timeout = null
        this.dispatcher.dispatch(this.opts, this)
      }, delay).unref()
    } else {
      // this.logger?.error({ err }, 'upstream request failed')
      this.handler.onError(err)
    }
  }
}

```

### PR DESCRIPTION
**Pull request by metcoder95**
## This relates to...
- Closes #2264

## Changes

- [x] Retry Agent
- [x] Documentation
- [x] Testing

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] Benchmarked (**optional**)
- [x] Documented
- [x] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by ronag** [comment]
Unfortunately I think retry is more complicated than this. See:

- https://github.com/nxtedition/nxt-lib/blob/e3ac11edbd265838cce537b0f4ab8ea298588f79/undici/interceptor/response-body-retry.js
- https://github.com/nxtedition/nxt-lib/blob/e3ac11edbd265838cce537b0f4ab8ea298588f79/undici/interceptor/response-retry.js
- https://github.com/nxtedition/nxt-lib/blob/e3ac11edbd265838cce537b0f4ab8ea298588f79/undici/interceptor/response-status-retry.js

File: lib/retry.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    // TODO: improve the erro handler for retry
+    if (
+      this.retryAfter == null ||
+      this.aborted ||
+      isDisturbed(this.opts.body)
+    ) {
+      return this.handler.onError(err)
+    }
+
+    this.timeout = setTimeout(() => {
```
Comment (ronag) [inline lib/retry.js:164]: This is only valid if we errored before forwarding any headers or body,

Comment (metcoder95) [reply lib/retry.js:164]: Or failed while connecting to the origin

**Comment by metcoder95** [comment]
Hey @ronag! Yes, I’m basing most of this work on your implementations, I just the work and opened the draft so I do not miss it and not make a mess with my local branches nor `stash` when switching between them. I’ll ping you once more when I have something more solid 😉

File: lib/retry.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
+const assert = require('node:assert')
+
+const createError = require('http-errors')
```
Comment (KhafraDev) [inline lib/retry.js:3]: not a fan, this adds 6 dependencies

Comment (metcoder95) [reply lib/retry.js:3]: Yeah, agree. I’d better create an util or something 👍

**Comment by kibertoad** [comment]
[mockttp](https://www.npmjs.com/package/mockttp) provides a very convenient way for testing weirder edge cases, such as connection lost or connection cancelled, highly recommend it.

You can look up the code examples here: https://github.com/kibertoad/undici-retry/blob/main/test/undiciRetry.spec.ts

File: lib/handler/RetryHandler.js [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+      maxTimeout: maxTimeout ?? 30 * 1000, // 30s,
+      timeout: minTimeout ?? 500, // .5s
+      timeoutFactor: timeoutFactor ?? 2,
+      max: max ?? 8,
+      // What errors we should retry
+      methods: methods ?? ['GET', 'HEAD', 'OPTIONS', 'PUT', 'DELETE', 'TRACE'],
+      // States weather or not retry on idempotent methods
+      idempotent: idempotent ?? false,
+      // Indicates which errors to retry
+      status: statusCodes ?? [500, 502, 503, 504],
```
Comment (kibertoad) [inline lib/handler/RetryHandler.js:44]: why no 429?

Comment (metcoder95) [reply lib/handler/RetryHandler.js:44]: I haven't implemented that one yet; as its nature is coming from a rate-limit logic, we need to handle its `Retry-After` header; I'm planning to add settings specifically to it and its own default logic

**Comment by metcoder95** [comment]
More testing and docs/types is missed, but should be good already for review 👍 

cc: @ronag

**Review by ronag — COMMENTED** [review]
Test and docs. LGTM otherwise.

**Review by ronag — APPROVED** [review]

File: lib/handler/RetryHandler.js [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+      this.resume = null
+
+      if (statusCode !== 206) {
+        return true
+      }
+
+      // TODO (fix): etag
+      // if (this.etag == null || this.etag !== etag) {
+      //   throw this.error
+      // }
```
[the remark this answers is not in the document]
Comment (metcoder95) [reply lib/handler/RetryHandler.js:189]: Missing feature, good catch; let me add it as well

File: lib/handler/RetryHandler.js [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+      if (statusCode !== 206) {
+        return true
+      }
+
+      let contentRange
+      if (
+        // Let's start with a weak etag check
+        (this.etag != null && this.etag !== headers.etag) ||
+        // If no content range
+        !(contentRange = parseRangeHeader(headers['content-range']))
```
Comment (ronag) [inline lib/handler/RetryHandler.js:191]: This is a bit too much syntax magic

File: lib/handler/RetryHandler.js [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        'ENOTFOUND',
+        'ENETDOWN',
+        'ENETUNREACH',
+        'EHOSTDOWN',
+        'EHOSTUNREACH',
+        'EPIPE'
+      ]
+    }
+
+    // TODO: hidde behind a symbol
```
Comment (ronag) [inline lib/handler/RetryHandler.js:67]: Just do it :)

File: lib/handler/RetryHandler.js [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+    const headers = parseHeaders(rawHeaders)
+    const err = new RequestRetryError(
+      'Request failed after retry',
+      statusCode,
+      {
+        headers,
+        count: this.count
+      }
+    )
```
Comment (ronag) [inline lib/handler/RetryHandler.js:169]: Only create error if failed?

File: lib/handler/RetryHandler.js [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+
+    this.count += 1
+
+    if (statusCode > 299) {
```
Comment (ronag) [inline lib/handler/RetryHandler.js:173]:
```suggestion
    if (statusCode >= 300) {
```

File: lib/handler/RetryHandler.js [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+      )
+    }
+
+    // Meant to just indicate weather or not trigger a retry
+    // It should return null to indicate we exhausted the retries
+    const retryAfter = this.retryOpts.retry(
+      err,
+      { counter: this.count, currentTimeout: this.retryAfter },
+      { retry: this.retryOpts, ...this.opts }
+    )
```
Comment (ronag) [inline lib/handler/RetryHandler.js:265]: I would like the retry method to be asynchronous and the timeout lives inside of it, i.e. instead of returning a duration.

Comment (ronag) [reply lib/handler/RetryHandler.js:265]: Either by returning a promise or taking a callback

File: lib/handler/RetryHandler.js [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        'EPIPE'
+      ]
+    }
+
+    // TODO: hidde behind a symbol
+    this.count = 0
+    this.timeout = null
+    this.retryAfter = null
+    this.head = 0
+    this.tail = null
```
Comment (ronag) [inline lib/handler/RetryHandler.js:72]: start and end is a bit clearer? Took me a bit to figure out what these were for. Maps better to existing vocabulary in regards to ranges.

File: types/retry-handler.d.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+     * It enables automatic retry between requets on idempotent methods.
+     * The method should be defined as safe accordingly to https://developer.mozilla.org/en-US/docs/Glossary/Safe/HTTP
+     *
+     * Note: It takes presedence over the `methods` option.
+     * 
+     * @type {boolean}
+     * @memberof RetryOptions
+     * @default false
+     */
+    idempotent?: boolean;
```
Comment (ronag) [inline types/retry-handler.d.ts:89]: Shouldn't this be true by default?

File: types/retry-handler.d.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * Error codes to be retried. e.g. `ECONNRESET`, `ENOTFOUND`, `ETIMEDOUT`, `ECONNREFUSED`, etc.
+     *
+     * @type {string[]}
+     */
+    codes?: string[];
```
Comment (ronag) [inline types/retry-handler.d.ts:95]: Missing default?

File: types/retry-handler.d.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * HTTP status codes to be retried.
+     *
+     * @type {number[]}
+     * @memberof RetryOptions
+     */
+    status?: number[];
```
Comment (ronag) [inline types/retry-handler.d.ts:102]: Missing default?

File: types/retry-handler.d.ts [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+     * @default true
+     */
+    retryAfter?: boolean;
+    /**
+     * HTTP methods to retry.
+     *
+     * @type {Dispatcher.HttpMethod[]}
+     * @memberof RetryOptions
+     */
+    methods?: Dispatcher.HttpMethod[];
```
Comment (ronag) [inline types/retry-handler.d.ts:78]: Missing default?

File: types/retry-handler.d.ts [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    retry?: RetryCallback;
+    /**
+     * Maximum number of retries to allow.
+     *
+     * @type {number}
+     * @memberof RetryOptions
+     * @default 5
+     */
+    max?: number;
```
Comment (ronag) [inline types/retry-handler.d.ts:39]:
```suggestion
    maxRetries?: number;
```

File: types/retry-handler.d.ts [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+  export interface RetryOptions {
+    /**
+     * Callback to be invoked on every retry iteration.
+     * It receives the error, current state of the retry object and the options object
+     * passed when instantiating the retry handler.
+     *
+     * @type {RetryCallback}
+     * @memberof RetryOptions
+     */
+    retry?: RetryCallback;
```
Comment (ronag) [inline types/retry-handler.d.ts:31]: All the options below are only relevant if `retry` is null. Maybe this is better represented as an union?

File: lib/handler/RetryHandler.js [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+
+function calculateRetryAfterHeader (retryAfter) {
+  const current = Date.now()
+  const diff = new Date(retryAfter).getTime() - current
+
+  return diff
+}
+
+class RetryHandler {
+  constructor (opts, handlers, retryOpts) {
```
Comment (ronag) [inline lib/handler/RetryHandler.js:20]: I would expect things to live inside opts, i.e. no retryOpts.

```suggestion
  constructor ({ retry: retryOpts, ...opts }, handlers) {
    if (typeof retryOpts === 'function') {
      retryOpts = { retry: retryOpts }
    } else if (typeof retryOpts === 'number') {
      retryOpts = { maxRetries: retryOpts }
    }
  }
```

Comment (ronag) [reply lib/handler/RetryHandler.js:20]: Of course this has the downside of potential conflict with other handlers... so maybe a separate handler opts is not bad either.

File: lib/handler/RetryHandler.js (hunk 6 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:67]: haha, left over, I forgot about that 😅

File: lib/handler/RetryHandler.js (hunk 9 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:265]: Hmm, has sense; it can enable several use cases

I'm imagine something like
```suggestion
    const retryAfter = this.retryOpts.retry(
      err,
      { counter: this.count, currentTimeout: this.retryAfter },
      { retry: this.retryOpts, ...this.opts }
    ).then(timeout => { /* timeout goes here */}, err => this.handler.onError(er))
```

Thoughts?

File: lib/handler/RetryHandler.js (hunk 10 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:72]: SGTM
I don't recall exactly why I went in that direction, I kind of recall because wanted to handle those two differently; but yeah your suggestion goes better 👍

File: types/retry-handler.d.ts (hunk 11 above)
Comment (metcoder95) [reply types/retry-handler.d.ts:89]: Have some conflicts deciding that; especially as the `methods` is also enabled by default, so it might make sense have one of them enabled. Maybe push for `idempotent` and disabled `methods`?

File: types/retry-handler.d.ts (hunk 12 above)
Comment (metcoder95) [reply types/retry-handler.d.ts:95]: Had hard times coping with adding the array for the JSDoc 😅 
But let me try once more

File: lib/handler/RetryHandler.js (hunk 17 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:20]: TBH, had the same thoughts as you, but the conflict won on my side so I decided to go with the weird opts. I'd prefer it to keep it there but maybe change the order first `opts` and later `handlers`?

Just wanted to avoid collisions, specially as the `opts` are reused between retries for the `dispatch`

File: lib/handler/RetryHandler.js (hunk 6 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:67]: Decided to not do it, do not see much value on doing so; implementation cannot access it, resolving for now

File: types/retry-handler.d.ts (hunk 16 above)
Comment (metcoder95) [reply types/retry-handler.d.ts:31]: Kind of, they `retryOpts` are also passed down to the custom `retry` fn; the goal was to allow them also make use of them if they want, so the settings they set, can also be consumed by the same implementation; I can see the potential but do not have strong opinion wether to keep it or not, thoughts?

File: lib/handler/RetryHandler.js (hunk 9 above)
Comment (ronag) [reply lib/handler/RetryHandler.js:265]: ACtually you don't need it to return any value, when the promise finishes then that's the retry.

Comment (metcoder95) [reply lib/handler/RetryHandler.js:265]: Yeah, I was actually about to ask you the same; basically the promise handles the retry; after it resolves we call `this.dispatch`, otherwise we handle the error

Comment (ronag) [reply lib/handler/RetryHandler.js:265]: TBH I would probably prefer a callback api, rather than promises, we don't use so much promises in our core code.

Comment (metcoder95) [reply lib/handler/RetryHandler.js:265]: Yeah, SGTM. Found some hiccups while refactoring to callback, I'll go through them later on but so far so good 👍

File: lib/handler/RetryHandler.js (hunk 17 above)
Comment (ronag) [reply lib/handler/RetryHandler.js:20]: Would you be fine with my suggestion?

File: types/retry-handler.d.ts (hunk 11 above)
Comment (ronag) [reply types/retry-handler.d.ts:89]: Hm, since there is overlap, maybe we can make it so methods is either an array of methods or the string `'idempotent'`?

Comment (ronag) [reply types/retry-handler.d.ts:89]: Not sure, maybe easiest to just remove this option?

File: lib/handler/RetryHandler.js [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+
+  onBodySent (chunk) {
+    return this.handler.onBodySent(chunk)
+  }
+
+  static [kRetryHandlerDefaultRetry] (err, { counter, currentTimeout }, opts) {
+    const { statusCode, code, headers } = err
+    const { method } = opts
+    const {
+      max,
```
Comment (ronag) [inline lib/handler/RetryHandler.js:103]:
```suggestion
      maxRetries,
```

File: lib/handler/RetryHandler.js [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+      timeout,
+      maxTimeout,
+      timeoutFactor,
+      status,
```
Comment (ronag) [inline lib/handler/RetryHandler.js:107]:
```suggestion
      statusCodes,
```

File: lib/handler/RetryHandler.js [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+      idempotent,
+      codes,
```
Comment (ronag) [inline lib/handler/RetryHandler.js:109]:
```suggestion
      errorCodes,
```

File: lib/handler/RetryHandler.js [hunk 21]
```diff
@@ -0,0 +1,347 @@
...
+  static [kRetryHandlerDefaultRetry] (err, { counter, currentTimeout }, opts) {
+    const { statusCode, code, headers } = err
+    const { method } = opts
+    const {
+      max,
+      timeout,
+      maxTimeout,
+      timeoutFactor,
+      status,
+      idempotent,
```
Comment (ronag) [inline lib/handler/RetryHandler.js:108]:
```suggestion
```

File: lib/handler/RetryHandler.js [hunk 22]
```diff
@@ -0,0 +1,347 @@
...
+    const err = new RequestRetryError(
+      'Request failed after retry',
+      statusCode,
+      {
+        headers,
+        count: this.count
+      }
+    )
+
+    this.count += 1
```
Comment (ronag) [inline lib/handler/RetryHandler.js:171]:
```suggestion
    this.retryCount += 1
```

File: lib/handler/RetryHandler.js [hunk 23]
```diff
@@ -0,0 +1,347 @@
...
+        !contentRange
+      ) {
+        this.abort(err)
+        return false
+      }
+
+      const { start, size, end = size } = contentRange
+
+      assert(this.head === start, 'content-range mismatch')
+      assert(this.tail == null || this.tail === end, 'content-range mismatch')
```
Comment (ronag) [inline lib/handler/RetryHandler.js:200]:
```suggestion
      assert(this.start === start, 'content-range mismatch')
      assert(this.end == null || this.end === end, 'content-range mismatch')
```

File: lib/handler/RetryHandler.js (hunk 17 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:20]: Sure thing 👍

File: lib/handler/RetryHandler.js (hunk 23 above)
Comment (metcoder95) [reply lib/handler/RetryHandler.js:200]: Weird, I remember doing those changes, maybe never pushed? 🤔

Comment (metcoder95) [reply lib/handler/RetryHandler.js]: Yeah, never pushed

File: types/retry-handler.d.ts (hunk 11 above)
Comment (metcoder95) [reply types/retry-handler.d.ts:90]: I'd say we better remove it, it will make our lifes and implementers easier. Anyways, if they want to automatically retry on certain method, they can set it in the `method` array

**Comment by metcoder95** [comment]
It seems its done, @ronag can you take a look?
Specially at the retry async function. I have some doubts about the signature, and possibly the handling. Open to suggestions 👍

File: docs/api/RetryHandler.md [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+
+Returns: `retryHandler`
+
+### Parameter: `Dispatch.DispatchOptions & RetryOptions`
+
+Extends: [`Dispatch.DispatchOptions`](Dispatcher.md#parameter-dispatchoptions).
+
+#### `RetryOptions`
+
+- **retry** `(err: Error, context: RetryContext, done: (result?: boolean) => void) => void` (optional) - Function to be called after every retry. It should call the `done` parameter to try the request once more.
```
Comment (ronag) [inline docs/api/RetryHandler.md:22]:
```suggestion
- **retry** `(err: Error, context: RetryContext, done: () => void) => void` (optional) - Function to be called after every retry. It should re-throw the error if no more retries should be performed.
```

File: docs/api/RetryHandler.md [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    retryOptions: {
+      // custom retry function
+      retry: function (err, state, done) {
+        counter++;
+
+        if (err.code && err.code === "UND_ERR_DESTROYED") {
+          return null;
+        }
+
+        if (err.statusCode === 206) return done(false);
```
Comment (ronag) [inline docs/api/RetryHandler.md:61]:
```suggestion
        if (err.statusCode === 206) throw err;
```

File: docs/api/RetryHandler.md [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+const handler = new RetryHandler(
+  {
+    ...dispatchOptions,
+    retryOptions: {
+      // custom retry function
+      retry: function (err, state, done) {
+        counter++;
+
+        if (err.code && err.code === "UND_ERR_DESTROYED") {
+          return null;
```
Comment (ronag) [inline docs/api/RetryHandler.md:58]:
```suggestion
          throw err;
```

File: lib/handler/RetryHandler.js [hunk 27]
```diff
@@ -0,0 +1,344 @@
...
+      code &&
+      code !== 'UND_ERR_REQ_RETRY' &&
+      code !== 'UND_ERR_SOCKET' &&
+      !errorCodes.includes(code)
+    ) {
+      return done(false)
+    }
+
+    // If a set of method are provided and the current method is not in the list
+    if (Array.isArray(methods) && !methods.includes(method)) return done(false)
```
Comment (ronag) [inline lib/handler/RetryHandler.js:129]:
```suggestion
    if (Array.isArray(methods) && !methods.includes(method)) {
      return done(false)
    }
```

I would hope we don't have these one-liners

File: lib/handler/RetryHandler.js [hunk 28]
```diff
@@ -0,0 +1,344 @@
...
+
+    // If a set of status code are provided and the current status code is not in the list
+    if (
+      statusCode != null &&
+      Array.isArray(statusCodes) &&
+      !statusCodes.includes(statusCode)
+    ) {
+      return done(false)
```
Comment (ronag) [inline lib/handler/RetryHandler.js:137]:
```suggestion
      throw err
```

File: docs/api/RetryHandler.md [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+
+```js
+const client = new Client(`http://localhost:${server.address().port}`);
+const chunks = [];
+const handler = new RetryHandler(
+  {
+    ...dispatchOptions,
+    retryOptions: {
+      // custom retry function
+      retry: function (err, state, done) {
```
Comment (ronag) [inline docs/api/RetryHandler.md:54]:
```suggestion
      retry: function (err, state, callback) {
```

File: docs/api/RetryHandler.md [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+        counter++;
+
+        if (err.code && err.code === "UND_ERR_DESTROYED") {
+          return null;
+        }
+
+        if (err.statusCode === 206) return done(false);
+        
+        setTimeout(done, 800);
```
Comment (ronag) [inline docs/api/RetryHandler.md:63]:
```suggestion
        setTimeout(() => done(null), 800);
```

File: lib/handler/RetryHandler.js [hunk 31]
```diff
@@ -0,0 +1,344 @@
...
+      currentTimeout != null && currentTimeout > 0 ? currentTimeout : timeout
+
+    // Any code that is not a Undici's originated and allowed to retry
+    if (
+      code &&
+      code !== 'UND_ERR_REQ_RETRY' &&
+      code !== 'UND_ERR_SOCKET' &&
+      !errorCodes.includes(code)
+    ) {
+      return done(false)
```
Comment (ronag) [inline lib/handler/RetryHandler.js:125]:
```suggestion
      return done(err)
```

File: lib/handler/RetryHandler.js [hunk 32]
```diff
@@ -0,0 +1,344 @@
...
+    if (
+      statusCode != null &&
+      Array.isArray(statusCodes) &&
+      !statusCodes.includes(statusCode)
+    ) {
+      return done(false)
+    }
+
+    // If we reached the max number of retries
+    if (counter > maxRetries) return done(false)
```
Comment (ronag) [inline lib/handler/RetryHandler.js:141]:
```suggestion
    if (counter > maxRetries) return done(err)
```

File: lib/handler/RetryHandler.js [hunk 33]
```diff
@@ -0,0 +1,344 @@
...
+    }
+
+    const retryTimeout =
+      retryAfterHeader > 0
+        ? Math.min(retryAfterHeader, maxTimeout)
+        : Math.min(currentTimeout * timeoutFactor ** counter, maxTimeout)
+
+    state.currentTimeout = retryTimeout
+
+    setTimeout(done, retryTimeout)
```
Comment (ronag) [inline lib/handler/RetryHandler.js:158]:
```suggestion
    setTimeout(() => done(null), retryTimeout)
```

File: lib/handler/RetryHandler.js [hunk 34]
```diff
@@ -0,0 +1,344 @@
...
+
+      const contentRange = parseRangeHeader(headers['content-range'])
+      if (
+        // Let's start with a weak etag check
+        (this.etag != null && this.etag !== headers.etag) ||
+        // If no content range
+        !contentRange
+      ) {
+        this.abort(
+          new RequestRetryError('Content-Range mismatch', statusCode, {
```
Comment (ronag) [inline lib/handler/RetryHandler.js:192]: Separate error for etag mismatch?

File: lib/handler/RetryHandler.js [hunk 35]
```diff
@@ -0,0 +1,344 @@
...
+    this.retryOpts.retry(
+      err,
+      {
+        state: { counter: this.retryCount++, currentTimeout: this.retryAfter },
+        opts: { retryOptions: this.retryOpts, ...this.opts }
+      },
+      onRetry.bind(this)
+    )
+
+    function onRetry (next) {
```
Comment (ronag) [inline lib/handler/RetryHandler.js:320]:
```suggestion
    function onRetry (err) {
```

File: lib/handler/RetryHandler.js [hunk 36]
```diff
@@ -0,0 +1,344 @@
...
+      if (next === false || this.aborted || isDisturbed(this.opts.body)) {
+        return this.handler.onError(err)
+      }
```
Comment (ronag) [inline lib/handler/RetryHandler.js:323]:
```suggestion
      if (err || this.aborted || isDisturbed(this.opts.body)) {
        return this.handler.onError(err)
      }
```

**Comment by metcoder95** [comment]
Applied on [1ed45f3](https://github.com/nodejs/undici/pull/2281/commits/1ed45f37b0cae527d49a86bba14407bb46ed6f7f) @ronag  👍

**Comment by ronag** [comment]
@metcoder95 are we ready here?

**Comment by metcoder95** [comment]
Yes 👍

**Comment by Ethan-Arrowood** [comment]
Nice!

**Comment by jiyuu-jin** [comment]
Is this exposed anywhere in the current NodeJS version?

**Comment by metcoder95** [comment]
This is exposed only through `undici`
